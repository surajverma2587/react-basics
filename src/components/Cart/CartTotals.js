import React from 'react';
import PropTypes from 'prop-types';

const CartTotals = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      CartTotals Component
    </div>
  );
};

/* display name for the component */
CartTotals.displayName = 'CartTotals';

CartTotals.propTypes = {
  /* add Proptypes here */
};

export default CartTotals;