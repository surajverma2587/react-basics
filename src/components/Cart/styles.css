.cart {
  margin-top: 20px;
}

.cart-item {
  padding-top: 10px;
  padding-bottom: 10px;
}