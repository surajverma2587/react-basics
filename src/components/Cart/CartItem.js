import React from 'react';
import PropTypes from 'prop-types';
import '../../assets/styles.css';

const CartItem = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      CartItem Component
    </div>
  );
};

/* display name for the component */
CartItem.displayName = 'CartItem';

CartItem.propTypes = {
  /* add Proptypes here */
};

export default CartItem;