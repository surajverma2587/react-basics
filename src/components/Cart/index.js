import React from 'react';
import PropTypes from 'prop-types';
import CartItem from './CartItem';
import CartTotals from './CartTotals';
import '../../assets/styles.css';

const renderCartItems = (/* cartItems */) => {
  // return list of CartItem components
}

const Cart = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* remember to call renderCartItems() to render a list of CartItem components */
  /* also render the CartTotal component and pass the required props */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      Cart Component
    </div>
  );
};

/* display name for the component */
Cart.displayName = 'Cart';

Cart.propTypes = {
  /* add Proptypes here */
};

export default Cart;