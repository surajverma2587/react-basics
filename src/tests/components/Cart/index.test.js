import React from 'react';
import { shallow } from 'enzyme';

import Cart from '../../../components/Cart';

const props = {
  subTotal: 10,
  taxes: 2,
  delivery: 3,
  total: 15,
  selectedItems: [
    {
      id: '1',
      title: 'title 1',
      description: 'description 1',
      price: 4
    },
    {
      id: '2',
      title: 'title 2',
      description: 'description 2',
      price: 10
    },
  ],
  onClickCartItem: jest.fn(),
};

describe('Cart Component', () => {
  describe('when props are valid', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Cart {...props} />);
    });
    
    it('should render snapshot of the component', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should render the component', () => {
      const cart = wrapper.find('.cart');
      expect(cart.exists()).toBe(true);
    });

    it('should render the CartItem components', () => {
      const cartItems = wrapper.find('CartItem');
      expect(cartItems.length).toEqual(2);
      cartItems.forEach((cartItem, index) => {
        expect(cartItem.prop('id')).toEqual(props.selectedItems[index].id);
        expect(cartItem.prop('title')).toEqual(props.selectedItems[index].title);
        expect(cartItem.prop('description')).toEqual(props.selectedItems[index].description);
        expect(cartItem.prop('price')).toEqual(props.selectedItems[index].price);
      });
    });

    it('should render the CartTotals component', () => {
      const cartTotals = wrapper.find('CartTotals');
      expect(cartTotals.exists()).toBe(true);
      expect(cartTotals.prop('subTotal')).toEqual(props.subTotal);
      expect(cartTotals.prop('taxes')).toEqual(props.taxes);
      expect(cartTotals.prop('delivery')).toEqual(props.delivery);
      expect(cartTotals.prop('total')).toEqual(props.total);
    });
  });
  
  describe('when props are invalid', () => {
    it('should not render the component when props are not present', () => {
      const wrapper = shallow(<Cart />);
      expect(wrapper.find('.cart').exists()).toBe(false);
    });
  
    it('should not render the component when any prop is invalid', () => {
      const invalidProps = [
        { ...props, subTotal: null },
        { ...props, taxes: null },
        { ...props, delivery: null },
        { ...props, total: null },
        { ...props, selectedItems: null },
        { ...props, selectedItems: [] },
        { ...props, onClickCartItem: null },        
      ];
  
      invalidProps.forEach(invalidProp => {
        const wrapper = shallow(<Cart {...invalidProp} />);
        expect(wrapper.find('.cart').exists()).toBe(false);
      });
    });
  });
});
