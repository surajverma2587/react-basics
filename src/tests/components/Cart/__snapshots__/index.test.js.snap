// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component when props are valid should render snapshot of the component 1`] = `
<div
  className="cart"
>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <span
        className="cart--title"
      >
        Cart
      </span>
      <hr />
      <CartItem
        description="description 1"
        id="1"
        key="1"
        onClick={[MockFunction]}
        price={4}
        title="title 1"
      />
      <CartItem
        description="description 2"
        id="2"
        key="2"
        onClick={[MockFunction]}
        price={10}
        title="title 2"
      />
      <CartTotals
        delivery={3}
        subTotal={10}
        taxes={2}
        total={15}
      />
    </div>
  </div>
</div>
`;
